[WIP] add optional boolean args "prepend" to cumsum, cumprod

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #128469
* __->__ #127675
* #128299

It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

davidberard98 committed:
[WIP] add optional boolean arg "true" to cumsum signature

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean arg "true" to cumsum signature"


TODO
* figure out how to do the AOTI stuff
* address any failures

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean arg "true" to cumsum signature"


TODO
* figure out how to do the AOTI stuff
* address any failures

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean arg "true" to cumsum signature"


TODO
* figure out how to do the AOTI stuff
* address any failures

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean arg "true" to cumsum signature"


TODO
* figure out how to do the AOTI stuff
* address any failures

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 commented:
@BowenBao @thiagocrepaldi could you recommend how I should update ONNX to handle this additional arg?

BowenBao commented:
cc @xadupre @titaiwangms for ONNX asks.

titaiwangms commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -787,12 +787,14 @@ def _reduce_with_dtype(onnx_op: str, name: str, allow_multi_dim_support: bool =
+@symbolic_helper.parse_args("v", "i", "none", "b")
 @_beartype.beartype
-def cumsum(g: jit_utils.GraphContext, input, dim, dtype):
+def cumsum(g: jit_utils.GraphContext, input, dim, dtype, full):
```
Is it because this one does not have default value as the other one, so it's caught by BC_linter?

titaiwangms commented:
For the test failure in `onnx/test_fx_to_onnx_with_onnxruntime.py` with the error message:

```
torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: Cannot find any perfect/nearest match of symbolic function for aten::cumsum.default,which should be registered under aten.cumsum.default.
```

Please xfail or skip them, as the symbolic function for onnx.dynamo_export is not yet in PyTorch.

cc @xadupre @justinchuby @shubhambhokare1

davidberard98 committed:
Update on "[WIP] add optional boolean arg "full" to cumsum signature"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning. This introduces an optional arg to specify `full=True` which will add the 0 at the beginning.

TODO
* figure out how to do the AOTI stuff
* ONNX - what should I update / not update
* TorchXLA - same ^^
* address any other failures
* choose a better name than "full" ?

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean arg "full" to cumsum signature"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning. This introduces an optional arg to specify `full=True` which will add the 0 at the beginning.

TODO
* figure out how to do the AOTI stuff
* ONNX - what should I update / not update
* TorchXLA - same ^^
* address any other failures
* choose a better name than "full" ?

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
rename 'full' -> 'prepend' on "[WIP] add optional boolean arg "full" to cumsum signature"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning. This introduces an optional arg to specify `full=True` which will add the 0 at the beginning.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58329265).

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58329265).

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

[ghstack-poisoned]

davidberard98 committed:
Update on "[WIP] add optional boolean args "prepend" to cumsum, cumprod"


It's frequently convenient to get a `cumsum` where the result is padded with a 0 at the beginning (or `cumprod` with a 1 padded at the beginning). This introduces an optional arg to specify `prepend=True` which will add the 0 at the beginning.

This PR only adds the signatures; it does not add the implementations. For the most part, this PR should add asserts if a user tries to call `cumsum` or `cumprod` with `prepend=True`. A follow-up PR will add the actual implementations.

TODO
* [ ] figure out how to do the AOTI stuff
* [x] ONNX - what should I update / not update
  - done I think
* [ ] TorchXLA - same ^^
* [ ] address any other failures
* [x] choose a better name than "full" ?
  - chose "prepend" (Ed's suggestion)
* [ ] .dimname versions

other TODO but for later PRs:
* refs / decompositions / lowerings
* common_methods_invocations

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

Differential Revision: [D58329265](https://our.internmc.facebook.com/intern/diff/D58329265)

[ghstack-poisoned]
